compiler/noirc_frontend/src/elaborator/types.rs

```diff
@@ -120,11 +137,31 @@ impl<'context> Elaborator<'context> {
                 );
             }
         }
+
+        // Check that any types with a type kind match the expected type kind supplied to this function
+        if let Type::NamedGeneric(_, name, resolved_kind) = &resolved_type {
+            if matches!(resolved_kind, TypeKind::Numeric) && matches!(kind, TypeKind::Normal) {
+                let expected_typ_err =
+                    CompilationError::ResolverError(ResolverError::NumericGenericUsedForType {
+                        name: name.to_string(),
+                        span: span.expect("Type should have span"),
+                    });
+                self.errors.push((expected_typ_err, self.file));
+                return Type::Error;
+            }
+        }
+
+        if resolved_type.is_nested_slice() {
+            self.push_err(ResolverError::NestedSlices {
+                span: span.expect("Type should have span"),
+            });
+        }
```

Comment:
Should be able to move this to only be checked in the Slice case

Reply:
What do you think about just moving it back to where it was in `resolve_type` after the the call to `resolve_type_inner` (https://github.com/noir-lang/noir/pull/5155#discussion_r1638370503)? Although rare, recursively checking the slice case would let us show the user in one run each place they have nested slices though.

Reply:
I moved it to the final check in `resolve_type` as we also want to account for nested slices inside of arrays and aliases.